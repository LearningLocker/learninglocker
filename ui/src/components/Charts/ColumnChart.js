import React from 'react';
import { BarChart as Chart, XAxis, YAxis, CartesianGrid } from 'recharts';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { AutoSizer } from 'react-virtualized';
import NoData from 'ui/components/Graphs/NoData';
import { compose } from 'recompose';
import uuid from 'uuid';
import {
  getResultsData,
  getShortModel,
  getChartData,
  hasData,
  renderTooltips,
  renderBars,
  renderLegend,
  hiddenSeriesState
} from './Chart';
import styles from './styles.css';

const sortData = data => data.sortBy(e => e.get('id'));
const getSortedData = results => labels => sortData(getResultsData(results)(labels));
const chartUuid = uuid.v4();
const renderBarChart = colors => (labels, toggleHiddenSeries, hiddenSeries) => stacked => data => ({ width, height }) => (
  /* eslint-disable react/no-danger */
  <div>
    <style
      dangerouslySetInnerHTML={{ __html: `
        .grid-${chartUuid} .recharts-cartesian-grid-vertical {
          visibility: hidden !important;
        }
      ` }} />
    <Chart
      className={`grid-${chartUuid}`}
      data={getChartData(data, hiddenSeries)}
      width={width}
      height={height}
      layout="horizontal">
      <XAxis type="category" dataKey="cellId" tickFormatter={getShortModel(data)} />
      <YAxis type="number" />
      <CartesianGrid strokeDasharray="1 1" />
      {renderLegend(labels, toggleHiddenSeries)}
      {renderBars(colors)(labels)(stacked)}
      {renderTooltips(data)}
    </Chart>
  </div>
  /* eslint-enable react/no-danger */
);
const renderChart = (model, component, axesLabels, chartWrapperFn) =>
(
  <div className={styles.chart}>
    <div className={`${styles.barContainer}`}>
      <span className={styles.yAxis}>
        {axesLabels.yLabel || model.getIn(['axesvalue', 'searchString'], 'Y-Axis')}
      </span>
      <div className={styles.chartWrapper}>
        {chartWrapperFn(component)}
      </div>
    </div>
    <div className={styles.xAxisLabel}>
      <span className={styles.xAxis}>
        {axesLabels.xLabel || model.getIn(['axesgroup', 'searchString'], 'X-Axis')}
      </span>
    </div>
  </div>
);
const renderChartResults = colors => (labels, toggleHiddenSeries, hiddenSeries) => stacked => results => (
  renderBarChart(colors)(labels, toggleHiddenSeries, hiddenSeries)(stacked)(getSortedData(results)(labels))
);
const renderResults = results => model => colors => (labels, toggleHiddenSeries, hiddenSeries) => stacked => axesLabels => chartWrapperFn => (
  renderChart(model, renderChartResults(colors)(labels, toggleHiddenSeries, hiddenSeries)(stacked)(results), axesLabels, chartWrapperFn)
);

export default compose(
  withStyles(styles),
  hiddenSeriesState
)(({
  results,
  labels,
  colors,
  stacked,
  axesLabels,
  chartWrapperFn = component => (<AutoSizer>{component}</AutoSizer>),
  toggleHiddenSeries,
  hiddenSeries,
  model
}) => (
  hasData(results) ? renderResults(results)(model)(colors)(labels, toggleHiddenSeries, hiddenSeries)(stacked)(axesLabels)(chartWrapperFn) : <NoData />
));
