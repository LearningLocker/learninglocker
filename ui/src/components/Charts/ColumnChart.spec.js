import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';
import ColumnChart from './ColumnChart';

describe('ColumnChartResult', () => {
  test('should render axes', () => {
    const mockResult = fromJS([[{
      '1970-01-01': {
        _id: '1970-01-01',
        count: 1,
        model: '1970-01-01'
      }
    }]]);

    const columnChart = renderer
      .create(
        <ColumnChart
          results={mockResult}
          labels={fromJS([undefined])}
          axesLabels={{
            yLabel: 'THE-Y',
            xLabel: 'THE-X'
          }}
          chartWrapperFn={() => (<div />)} />
      )
      .toJSON();

    expect(columnChart).toMatchSnapshot();
  });
});
