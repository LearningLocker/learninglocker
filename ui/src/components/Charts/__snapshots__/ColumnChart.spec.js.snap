// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnChartResult should render axes 1`] = `
<div
  className="chart"
>
  <div
    className="barContainer"
  >
    <span
      className="yAxis"
    >
      THE-Y
    </span>
    <div
      className="chartWrapper"
    >
      <div />
    </div>
  </div>
  <div
    className="xAxisLabel"
  >
    <span
      className="xAxis"
    >
      THE-X
    </span>
  </div>
</div>
`;
